menelaus/injection/class_manipulation.py

```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
+    Accepts ``pandas.DataFrame`` with column names or ``numpy.ndarray``
+    with column indices.
+
+    Ref. :cite:t:`LTFmethods`
+
+    Args:
+        data (np.array): data to inject with drift
+        target_col (int or str): column index/label of targets column
+        from_index: row index at which to start class swap
+        to_index: row index at which to end (non-inclusive) class swap
+        classes (list[int]): list of classes with altered sampling chance
+        p_classes (float): altered sampling chance for members of ``classes``
+
+    Returns:
+        np.array or pd.DataFrame: copy of data, resampled with shifted
+            class probability for 1 or more desired classes
+    """
+    ret = np.copy(data)
+    classes = np.unique(ret[:, target_col])
+
+    # distribution for each data point, and reordering of each point by class
+    p_distribution = np.array([])
+    sample_grouped = np.array([])
+
+    # locate all classes to shift
+    shifted_classes_idx = np.where(data[:, target_col].isin(classes))[0]
+
+    # locate each class in window
+    for cls in classes:
+        cls_idx = np.where(data[:, target_col] == cls)[0]
+        cls_idx = cls_idx[
+            (cls_idx < to_index) &
+            (cls_idx >= from_index)
+        ]
+
+        if cls in classes:
+            # pts from shifted classes drawn with separate probability
+            p_individual = p_classes / shifted_classes_idx.shape[0]
+        else:
+            # pts from remaining classes drawn with uniform probability
+            p_individual = ((1 - p_classes) / (classes.shape[0] - 1)) / cls_idx.shape[0]
+
+        # append to grouped array and corresponding distribution
+        sample_grouped = np.concatenate((sample_grouped, data[cls_idx]))
+        p_distribution = np.concatenate(p_distribution, np.ones(cls_idx.shape[0]) * p_individual)
+
+    # shuffled sample over window, with replacement, with weights
+    sample = np.random.choice(sample_grouped, to_index-from_index, True, p_distribution)
+    ret[from_index:to_index] = sample
+    return ret
+
+
+def class_dirichlet_shift(data, target_col, from_index, to_index, alpha):
+    """ 
+    Resamples the data over a specified window, per a generated Dirichlet
+    distribution (with specified alpha) over all labels.
+    Accepts ``pandas.DataFrame`` with column names or ``numpy.ndarray``
+    with column indices.
+
+    Ref. :cite:t:`LTFmethods`
+
+    Args:
+        data (np.array): data to inject with drift
+        target_col (int or str): column index/label of targets column
+        from_index: row index at which to start class swap
+        to_index: row index at which to end (non-inclusive) class swap
+        alpha (dict): used to derive alpha parameter for Dirichlet
+            distribution. Keys are all labels, values are the desired
+            average weight (typically ``int``) per label when resampling. 
+        p_classes (float): altered sampling chance for members of ``classes``
```

Comment:
I'm assuming p_classes is in the documentation from copying it over, it's not a paramater right?

Reply:
Good catch, thanks! Not a parameter for D-shift